rl4co/models/zoo/deepaco/antsystem.py

```diff
@@ -0,0 +1,245 @@
+            log_p, mask = NARDecoder._get_log_p(td, heatmaps_logp, num_starts)
+            td = self.decode_strategy.step(log_p, mask, td)
+            td = env.step(td)["next"]
+
+        outputs, actions, td, env = self.decode_strategy.post_decoder_hook(td, env)
+        reward = env.get_reward(td, actions)
+
+        if self.require_logp:
+            self.all_records.append((outputs, actions, reward, td.get("mask", None)))
+
+        # reshape from (batch_size * n_ants, ...) to (batch_size, n_ants, ...)
+        reward = reward.view(self.batch_size, self.n_ants)
+        actions = actions.view(self.batch_size, self.n_ants, -1)
+
+        return actions, reward
+
+    def _local_search(self, actions: Tensor, reward: Tensor) -> Tuple[Tensor, Tensor]:
+        """Perform local search on the actions and reward obtained.
+
+        This method is reserved for extensions and can be overridden in child classes to implement
+        specific local search strategies. By default, it returns the original actions and reward
+        without any modifications.
+
+        Args:
+            actions: Actions chosen by the algorithm.
+            reward: Reward obtained after the sampling step.
+
+        Returns:
+            actions: The modified actions
+            reward: The modified reward
+
+        """
+        return actions, reward
+
+    def _update_results(self, actions, reward):
+        # update the best-trails recorded in self.final_actions
+        best_index = reward.argmax(-1)
+        best_reward = reward[self._batchindex, best_index]
+        best_actions = actions[self._batchindex, best_index]
+
+        if self.final_actions is None or self.final_reward is None:
+            self.final_actions = best_actions.clone()
+            self.final_reward = best_reward.clone()
+        else:
+            require_update = self._batchindex[self.final_reward <= best_reward]
+            self.final_actions[require_update, :] = best_actions[require_update]
+            self.final_reward[require_update] = best_reward[require_update]
+
+        return best_index
+
+    def _update_pheromone(self, actions, reward):
+        # calculate Δphe
+        delta_pheromone = torch.zeros_like(self.pheromone)
+        from_node = actions
+        to_node = torch.roll(from_node, 1, -1)
+        mapped_reward = self._reward_map(reward).detach()
+        batch_action_indices = self._batch_action_indices(
+            self.batch_size, actions.shape[-1], reward.device
+        )
+
+        for ant_index in range(self.n_ants):
+            delta_pheromone[
+                batch_action_indices,
+                from_node[:, ant_index].flatten(),
+                to_node[:, ant_index].flatten(),
+            ] += mapped_reward[batch_action_indices, ant_index]
+
+        # decay & update
+        self.pheromone *= self.decay
+        self.pheromone += delta_pheromone
+
+    def _reward_map(self, x):
+        # map reward from $\mathbb{R}$ to $\mathbb{R}^+$
```

Comment:
[Minor, no need to modify now] For documentation purposes in ReadTheDocs, math can be done

```python
   def _reward_map(self, x):
        """Map reward :math:`f: \mathbb{R} \rightarrow \mathbb{R}^+`"""
```